### PR TITLE
[ENH] Exogenous Handling in `TinyTimeMixer` Forecaster

### LINKED ISSUE #8629: [ENH] exogenous capability for TinyTimeMixerForecaster
According to https://github.com/sktime/sktime/issues/8522, TinyTimeMixer is capable of handling exogenous data - we should add that feature. See #8522 for discussion how it could be acced.

### PR DESCRIPTION
**Pull request by akshathmangudi**
#### Reference Issues/PRs
Fixes #8629 

The PR implements native exogenous variables support for `TinyTimeMixerForecaster`, exposing the already available capabilities of the underlying `TinyTimeMixer`model through the sktime interface. Previously, TTM was configured to ignore exogenous variables through the flag variable `"ignores-exogeneous-X": True` despite the underlying models supporting them. 

#### What does this implement/fix? Explain your changes.
Changes made: 
- **Updated tag**: `"ignores-exogeneous-X": False` to show exogenous support. 
- Added `X` parameter with multi-index support. 
- Updated `_fit` and `_predict` to handle exogenous variables.

#### Did you add any tests for the change?
Yes. I noticed that TTM does not have a dedicated test file inside `sktime/forecaster/tests/` so I have created a minimal test file with the following methods and their functions: 

`test_ttm_exogenous_basic()`: Tests the core exogenous variables functionality. 
- Uses `load_longley()` which provides both target series `y` and exogenous variables `X`. 
- Calls the updated forecaster `forecaster.fit(y, X=X, fh=[1,2])` to verify the model can be trained with exogenous variables. 
- Asserts that the model predictions are successfully generated. 

`test_ttm_exogenous_variables_validation_split()`: Tests that validation split works correctly with exogenous variables
- Sets `validation_split=0.2` to enable train/validation splitting.
- Verifies that both `y` and `X` are properly split using `temporal_train_test_split`.
- Asserts that the validation split doesn't break exogenous variable handling.

`test_ttm_exogenous_variables_backward_compatibility()`: Critical test - ensures existing code continues to work without changes
- Loads only target: Uses `y, _ = load_longley()` (discards X, only uses target series)
- Traditional usage: Calls `forecaster.fit(y, fh=[1, 2]` without the `X` parameter.
- Asserts that the old usage pattern still works exactly as before.

#### Any other comments?
**Exact diff change:**
```python
diff --git a/sktime/forecasting/ttm.py b/sktime/forecasting/ttm.py
index 97488ccb8..6d5db7741 100644
--- a/sktime/forecasting/ttm.py
+++ b/sktime/forecasting/ttm.py
@@ -158,6 +158,17 @@ class TinyTimeMixerForecaster(_BaseGlobalForecaster):
           performance but requires more computational power and time. Allows
           model path to be None.
 
+    *Exogenous Variables Support*
+
+    TTM supports exogenous variables (external factors) that can improve forecasting
+    accuracy. The model accepts exogenous variables that are known for both the
+    historical period and the future forecasting horizon.
+
+    When using exogenous variables:
+    - The X parameter should contain exogenous data covering both past and future periods
```

### CODE REVIEW

File: sktime/forecasting/ttm.py [hunk 1]
```diff
@@ -470,6 +547,9 @@ def _fit(self, y, X, fh):
         # Get the model
         self.model = trainer.model
 
+        # Store exogenous variables info for prediction
+        self._X = X
```
Comment (julian-fong) [inline sktime/forecasting/ttm.py]: it doesn't look like this is being used anywhere else? Do we need it for something?

File: sktime/forecasting/ttm.py [hunk 2]
```diff
@@ -158,6 +158,18 @@ class TinyTimeMixerForecaster(_BaseGlobalForecaster):
           performance but requires more computational power and time. Allows
           model path to be *None*.
 
+    **Exogenous Variables Support**
```
Comment (julian-fong) [inline sktime/forecasting/ttm.py]: This section should be part of the description and not under the Parameters section

File: sktime/forecasting/ttm.py [hunk 3]
```diff
@@ -716,16 +830,103 @@ def __getitem__(self, i):
...
         n = i // self.single_length
 
-        past_values = self.y[n, m : m + self.context_length, :]
-        future_values = self.y[
-            n,
-            m + self.context_length : m + self.context_length + self.prediction_length,
-            :,
-        ]
+        # Handle edge case where data is shorter than context + prediction length
+        if self.n_timestamps < self.context_length + self.prediction_length:
```
Comment (julian-fong) [inline sktime/forecasting/ttm.py:807]: Can you explain what this edge case is for and how one might encounter it using an example? `self.n_timestamps < self.context_length + self.prediction_length` would mean we have zero actual samples in the dataset

File: sktime/forecasting/ttm.py [hunk 4]
```diff
@@ -716,16 +830,103 @@ def __getitem__(self, i):
...
 
-        return {
+        result = {
             "past_values": tensor(past_values).float(),
             "observed_mask": tensor(observed_mask).float(),
             "future_values": tensor(future_values).float(),
         }
+
+        # Add exogenous variables if available
+        if self.X is not None:
```
Comment (julian-fong) [inline sktime/forecasting/ttm.py:859]: if you're padding for `self.X` would you also need to pass some sort of mask as well? not 100% sure

**Comment by julian-fong** [comment]
don't worry about the failures on mac-os they're unrelated. fix is on another PR atm.

also please copy your previous description in the closed pr and add it here - you can leave your current description at the top no need to remove

File: sktime/forecasting/ttm.py (hunk 1 above)
Comment (akshathmangudi) [reply sktime/forecasting/ttm.py]: good call! removing it now.

**Comment by akshathmangudi** [comment]
here was the previous description: 


#### Reference Issues/PRs
Fixes: resolves issue #8629 

The PR implements native exogenous variables support for `TinyTimeMixerForecaster`, exposing the already available capabilities of the underlying `TinyTimeMixer`model through the sktime interface. Previously, TTM was configured to ignore exogenous variables through the flag variable `"ignores-exogeneous-X": True` despite the underlying models supporting them. 

#### What does this implement/fix? Explain your changes.
Changes made: 
- **Updated tag**: `"ignores-exogeneous-X": False` to show exogenous support. 
- Added `X` parameter with multi-index support. 
- Updated `_fit` and `_predict` to handle exogenous variables.

#### Did you add any tests for the change?
Yes. I noticed that TTM does not have a dedicated test file inside `sktime/forecaster/tests/` so I have created a minimal test file with the following methods and their functions: 

`test_ttm_exogenous_basic()`: Tests the core exogenous variables functionality. 
- Uses `load_longley()` which provides both target series `y` and exogenous variables `X`. 
- Calls the updated forecaster `forecaster.fit(y, X=X, fh=[1,2])` to verify the model can be trained with exogenous variables. 
- Asserts that the model predictions are successfully generated. 

`test_ttm_exogenous_variables_validation_split()`: Tests that validation split works correctly with exogenous variables
- Sets `validation_split=0.2` to enable train/validation splitting.
- Verifies that both `y` and `X` are properly split using `temporal_train_test_split`.
- Asserts that the validation split doesn't break exogenous variable handling.

`test_ttm_exogenous_variables_backward_compatibility()`: Critical test - ensures existing code continues to work without changes
- Loads only target: Uses `y, _ = load_longley()` (discards X, only uses target series)
- Traditional usage: Calls `forecaster.fit(y, fh=[1, 2]` without the `X` parameter.
- Asserts that the old usage pattern still works exactly as before.

#### Any other comments?
**Exact diff change:**
```python
diff --git a/sktime/forecasting/ttm.py b/sktime/forecasting/ttm.py
index 97488ccb8..6d5db7741 100644
--- a/sktime/forecasting/ttm.py
+++ b/sktime/forecasting/ttm.py
@@ -158,6 +158,17 @@ class TinyTimeMixerForecaster(_BaseGlobalForecaster):
           performance but requires more computational power and time. Allows
           model path to be None.
 
+    *Exogenous Variables Support*
+
+    TTM supports exogenous variables (external factors) that can improve forecasting
+    accuracy. The model accepts exogenous variables that are known for both the
+    historical period and the future forecasting horizon.
+
+    When using exogenous variables:
+    - The X parameter should contain exogenous data covering both past and future periods
```

File: sktime/forecasting/ttm.py (hunk 3 above)
Comment (akshathmangudi) [reply sktime/forecasting/ttm.py:807]: so, imagine if we have the context of about 10 past steps, and we would to predict the next 3 values BUT we only have about 8 timestamps to work with, then this will handle the situations where the available time series data is shorter than the model needs to train. 

does that make sense?

File: sktime/forecasting/ttm.py (hunk 4 above)
Comment (akshathmangudi) [reply sktime/forecasting/ttm.py:859]: no i don't think so. i believe the implementation already treats these exogenous variables as known future information. i think we should be good.

**Comment by akshathmangudi** [comment]
@julian-fong can you please review my changes?

**Comment by akshathmangudi** [comment]
image

@fkiraly how would you like me to approach this? I believe it's because we did not support exogenous variables for this forecaster before (not natively) so we expected nothing, but now that we're getting some output, would we want to change the expected output such that it matches with the doctest?

**Comment by fkiraly** [comment]
this is checking the printout, it is how doctest works.

* the line below the `fit` should be `TinyTimeMixerForecaster(...)` (as written)
* the dict also seems to be printed out. You might have print statements somewhere in the code?

**Comment by akshathmangudi** [comment]
im not able to isolate this issue. ive been looking at it for the past hour, there are no print or return statements that might print out the dictionary either from the wrapper or the base class. i don't see any overrides either.

**Comment by fkiraly** [comment]
I would suggest, check when you execute the code in the doctest, line by line.

Also, add the expected output after the `fit` line, and compare

**Comment by akshathmangudi** [comment]
the problem with adding the expected output is that the results are not reproducible i.e. the seed isn't set, so everytime i run the doctest, we're getting different loss values, training time values, etc. 

i did try adding `# doctest: +SKIP` though and though we skip the test, im not getting any errors.

**Comment by fkiraly** [comment]
There is clearly an unexpected printout, but I was unable to find the bug. I have opened an issue instead: https://github.com/sktime/sktime/issues/8732
